lib/vanagon/component/source/git.rb

```diff
@@ -1,37 +1,88 @@
 require 'vanagon/utilities'
+require 'vanagon/errors'
+# This stupid library requires a capital 'E' in its name
+# but it provides a wealth of useful constants
+require 'english'
 
 class Vanagon
   class Component
     class Source
       class Git
-        include Vanagon::Utilities
-        attr_accessor :url, :ref, :workdir, :version, :cleanup
+        attr_accessor :url, :ref, :workdir, :clone_depth
+        attr_reader :default_options, :version
+
+        GIT = Vanagon::Utilities.which('git')
+
+        class << self
+          # Simple wrapper around git command line executes the given commands and
+          # returns the results. Required for Git.valid_remote? to function. Wrapped
+          # by a private instance method of the same name, with the same options.
+          #
+          # @param command_string [String] The commands to be run
+          # @param dir [String,Pathname] Directory to run the commands in
+          # @param raisable [boolean] if this function should raise an error on a git failure
+          # @param quiet [boolean] if this function should return output as a String or Boolean status
+          # @return [String] the output of the command (if :quiet is false)
+          # @return [Boolean] whether or not the command executed successfully (if :quiet is true)
+          def git(command_string, dir: Dir.pwd, raisable: false, quiet: false)
+            output = Dir.chdir(dir) { %x(#{GIT} #{command_string}) }
+
+            if raisable && !$CHILD_STATUS.success?
+              raise Vanagon::GitError, %('git #{command_string}' failed:\n#{output})
+            end
+
+            return $CHILD_STATUS.success? if quiet
+
+            output
+          end
+          private :git
+
+          # Attempt to connect to whatever URL is provided and
+          # return True or False depending on whether or not
+          # `git` thinks it's a valid Git repo.
+          #
+          # @return [Boolean] whether #url is a valid Git repo or not
+          def valid_remote?(url)
```

Comment:
Ostensibly, and @shrug and I evaluated using something like rugged (Ruby bindings to libgit2) for these operations. The benefit was negligible for our use case, since git cloning is basically analogous to an HTTP `GET` or a local file `cp` for how we're using Git.

We could probably research other options down the road.